Project Status Page (5.4 branch)

dwelch2344 opened the pull request:
With Protractor becoming LTS and given the partnership with [HeroDevs](https://www.herodevs.com/) to provide support past August 2023, we had discussed adding documentation to explain what the future holds.

The bulk of this is adding a `banner` to every page and a `status page` that gives more information about EOL. Both of these have 3 minor variants of verbiage, depending on the browser's date on render. Screenshots below:

## Now - Aug 2023
![current](https://user-images.githubusercontent.com/160142/233693065-4245924c-f73f-40ad-aecf-d5cdae2fddff.png)

## Sept 2023 - Aug 2024
![sept 2023 to aug 2024](https://user-images.githubusercontent.com/160142/233693080-f92d9be7-9607-4965-a517-00ff1a0f9dfe.png)

## Sept 2024+
![sept 2024](https://user-images.githubusercontent.com/160142/233693107-53bc9dce-2586-47a0-8158-84102bc014da.png)

_Note_: This PR replaces #5557 (and #5556) to both address EOL and update the docs / builds for the website. Apologies for another PR, but it looks like the current site is currently running off a 5.4 commit that never made it to the release branch.

dwelch2344 committed:
docs(website) clean up docs & add status page + banner

mgechev commented on lib/selenium-webdriver/webdriver.js:
```diff
@@ -236,7 +236,7 @@ webdriver.WebDriver.prototype.executeAsyncScript = (script, var_args) => {};
  * browser.call(logText, counter);
  *
- * @param {function(...): (T|promise.Promise<T>)} fn The function to
+ * @param {function(...*): (T|promise.Promise<T>)} fn The function to
```
If this change is not related to the banner update, I'd recommend reverting back. I assume you're not able to build without it?

dwelch2344 replied:
Happy to do so! Yeah, I'm a little concerned about the build process to actually deploy this as there were a few docs / build scripts that felt like they may be half-updated. I was able to get it to build, but it required updating the `dgeni` version for a bug fix (and thus where this came in)

dgp1130 commented on website/README.md:
```diff
@@ -6,21 +6,39 @@ Generates the documentation for the protractor website.
 cd website/
 npm install
-npm run build
+npx gulp
```
Can you include the version of `gulp` you're using for future reference? That way updates to `gulp` don't break this process.

dwelch2344 replied:
I believe `npx gulp` will first look in the `node_modules/.bin/` first, so it should pin directly to whatever's listed in the `package.json`. That said, happy to include the specific version if preferable!

dgp1130 replied:
Oh I didn't realize that was properly listed as a dev dependency. How about we make `"gulp": "gulp"` a script and then document it as `npm run gulp`. That way it's specifically tied to the installed dependency, not downloaded over the network (ex. if the user forgot to `npm install` first).

dwelch2344 replied:
Great call! I had totally missed it, but `npm run build` would essentially do just that

dgp1130 commented on website/README.md:
````diff
@@ -6,21 +6,39 @@ Generates the documentation for the protractor website.
+#FailureMessage Object: 0x306a626c0
+```
+
+Eventually the project should migrate to Gulp 5 / Node 18+ to solve this issue, but in the meantime a simple workaround to run `npm run arm64_repair` (after every `npm install` in the `website/` directory)
````
Should this be included in the "How to run" instructions above?

dgp1130 commented on website/partials/status.html:
```diff
@@ -0,0 +1,40 @@
+  </div>
+  
+  <h2>What does End Of Life mean?</h2>
+  The packages will remain accessible on GitHub, npm, Bower, and Release archive. This website will also remain here indefinitely. The project will be archived on GitHub and remain accessible indefinitely, but will no longer be open to issues / comments / etc. 
```
> This website will also remain here indefinitely.

@mgechev, are we comfortable committing to the website staying up _forever_? I doubt GitHub Pages would push a significant breaking change here, but if they did and it was a non-trivial effort to move off it, I expect we probably wouldn't bother given Internet Archive as the ultimate workaround.

mgechev replied:
Yeah, I agree. Let's change to "foreseeable future"

We're not anticipating making the website unavailable, but inevitably it'll happen one day.

dgp1130 reviewed (changes requested):
I recommend squashing these commits, as the last two don't provide a lot of value and just fixup stuff from the first.

Wording of the messages sounds good to me, just one note about the website lifetime.

dwelch2344 commented:
@mgechev @dgp1130 Thanks for the review! Cleaned up 3 of the 4 comments, left a question on the final one for @dgp1130

dwelch2344 commented:
@dgp1130 I think we're good to go. I squashed to a single commit and cleaned up the issues, but definitely let me know if  there's anything else we can do on our end! Thanks

dgp1130 reviewed (approved):
Yep, looks great. Thanks @dwelch2344!

dgp1130 commented:
I'll take a look at building and pushing to `gh-pages` tomorrow to see if we can get this out.

dgp1130 commented:
Sorry, got distracted with v16. Pushed the release now and protractortest.org seems to be updated correctly.

I needed to add back in that typing change to get it to build, not sure how you were avoiding that. I also needed to swap a boolean condition as it seems the header message was displayed in the wrong order.

https://github.com/angular/protractor/commit/bb0e475b449fe2ed2a71c2e63bc1ab46f7a03309
